mrmustard/lab_dev/samplers.py

```diff
@@ -0,0 +1,207 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+from itertools import product
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Sequence
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number, QuadratureEigenstate
+from .circuit_components import CircuitComponent
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler(ABC):
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        povms: The POVMs of this sampler.
+    """
+
+    def __init__(
+        self,
+        meas_outcomes: Sequence[Any],
+        povms: CircuitComponent | Sequence[CircuitComponent],
+    ):
+        self._povms = povms
+        self._meas_outcomes = meas_outcomes
+
+        self._outcome_arg = None
+
+    @property
+    def povms(self) -> CircuitComponent | Sequence[CircuitComponent]:
+        r"""
+        The POVMs of this sampler.
+        """
+        return self._povms
+
+    @property
+    def meas_outcomes(self) -> Sequence[Any]:
+        r"""
+        The measurement outcomes of this sampler.
+        """
+        return self._meas_outcomes
+
+    @abstractmethod
+    def probabilities(self, state: State, atol: float = 1e-4) -> Sequence[float]:
+        r"""
+        Returns the probability distribution of a state w.r.t. measurement outcomes.
+
+        Args:
+            state: The state to generate the probability distribution of. Note: the
+                input state must be normalized.
+            atol: The absolute tolerance used for validating that the computed
+                probability distribution sums to ``1``.
+        """
+
+    def sample(self, state: State, n_samples: int = 1000, seed: int | None = None) -> np.ndarray:
+        r"""
+        Returns an array of samples given a state.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+
+        Returns:
+            An array of samples such that the shape is ``(n_samples, n_modes)``.
+        """
+        initial_mode = state.modes[0]
+        initial_samples = self.sample_prob_dist(state[initial_mode], n_samples, seed)
+
+        if len(state.modes) == 1:
+            return initial_samples
+
+        unique_samples, counts = np.unique(initial_samples, return_counts=True)
+        ret = []
+        for unique_sample, counts in zip(unique_samples, counts):
+            meas_op = self._get_povm(unique_sample, initial_mode).dual
+            reduced_state = (state >> meas_op).normalize()
```

Comment:
I think this line can be more efficient in subclasses. Here it computes an inner product between the state and the POVM element and then calls `normalize` which computes more inner products. For a Homodyne measurement this is equivalent to calling the Ansatz in the Q basis with the variable value equal to the measurement outcome, and then dividing by the probability of the outcome (both of which are already known at this point). For a PNR measurement we have to go though the polynomial, but still it must be more efficient than creating several `Number` states and computing inner products against them.